oracle/src/main.rs

```diff
@@ -79,55 +197,46 @@ async fn main() -> Result<(), Error> {
 
-    let interval = Duration::from_millis(opts.interval_ms);
-    let exchange_rate = FixedU128::checked_from_integer(opts.exchange_rate).ok_or(Error::InvalidExchangeRate)?;
-
-    let conversion_factor = FixedU128::checked_from_rational(
-        10_u128.pow(opts.collateral_decimals),
-        10_u128.pow(opts.wrapped_decimals),
-    )
-    .unwrap();
+    let blockstream_url = if let Some(mut url) = opts.blockstream {
+        url.set_path(&format!("{}/fee-estimates", url.path()));
+        Some(url)
+    } else {
+        None
+    };
+
+    // TODO: configurable collateral currencies
+    let coingecko_url = if let Some(mut url) = opts.coingecko {
+        url.set_path(&format!("{}/simple/price", url.path()));
+        url.set_query(Some("ids=bitcoin&vs_currencies=dot"));
+        Some(url)
+    } else {
+        None
+    };
+
+    let conversion_factor =
+        FixedU128::checked_from_rational(10_u128.pow(DOT_DECIMALS), 10_u128.pow(BTC_DECIMALS)).unwrap();
 
     loop {
-        let exchange_rate = if opts.coingecko {
-            match get_exchange_rate_from_coingecko().await {
-                Ok(exchange_rate) => {
-                    // exchange_rate given in BTC/DOT so there is no need to adjust
-                    FixedU128::checked_from_integer(exchange_rate).unwrap()
-                }
-                Err(err) => {
-                    error!("Could not get exchange rate from CoinGecko: {}", err);
-                    sleep(ERR_RETRY_WAIT).await;
-                    continue;
-                }
-            }
-        } else {
-            exchange_rate
-        };
-
-        let exchange_rate = exchange_rate
-            .checked_mul(&conversion_factor)
-            .ok_or(Error::InvalidExchangeRate)?;
-
-        info!(
-            "Setting exchange rate: {} ({})",
-            exchange_rate,
-            chrono::offset::Local::now()
+        let parachain_rpc =
+            InterBtcParachain::from_url_with_retry(&opts.btc_parachain_url, signer.clone(), opts.connection_timeout_ms)
+                .await?;
+
+        let (left, right) = join!(
+            do_and_report(submit_bitcoin_fees(
+                &parachain_rpc,
+                UrlOrDefault::from_args(blockstream_url.clone(), opts.bitcoin_fee)
+            )),
+            do_and_report(submit_exchange_rate(
+                &parachain_rpc,
+                UrlOrDefault::from_args(coingecko_url.clone(), opts.exchange_rate),
+                conversion_factor
+            ))
         );
 
-        let result = InterBtcParachain::from_url_with_retry(
-            &opts.btc_parachain_url.clone(),
-            signer.clone(),
-            Duration::from_millis(opts.connection_timeout_ms),
-        )
-        .await?
-        .set_exchange_rate(exchange_rate)
-        .await;
-
-        if let Err(e) = result {
-            error!("Error: {}", e.to_string());
+        if left.is_err() || right.is_err() {
+            return Err(Error::Shutdown);
```

Comment:
Why exit on failure? Imo the oracle should never quit - it should keep running forever, without depending on external restarts

Reply:
This will only return after the `backoff` period has expired, in which case something is seriously wrong (can't reach CoinGecko / Blockstream) and we should shut down.

Reply:
Just pushed a new commit, I actually saw no reason not to use `backoff` here instead and then error.

Reply:
I still think continuing to retry is better, but it doesn't have to block this pr..

Reply:
Moved to https://github.com/interlay/interbtc-clients/issues/164